File: lab07/log_probability.py
#!/usr/bin/python
import sys
import re
import glob
import math

# FUNCTION DEFINITIONS
def findWordFrequency(totalWords, totalWordMatches):
    return float(totalWordMatches) / float(totalWords)

def findTotal(filename):
    fileHandle = open(filename, 'r') # open file for reading
    total = 0
    for line in fileHandle:
        words = re.split("[^a-zA-Z]", line)
        for word in words:
            if not word == "":
                total += 1
    fileHandle.close()
    return total

def findWordMatches(filename, wordToMatch):
    total = 0
    fileHandle = open(filename, 'r')
    wordToMatch = wordToMatch.lower()
    for line in fileHandle:
        words = re.split(r"[^a-zA-Z]", line)
        for word in words:
            word = word.lower()
            if word == wordToMatch:
                total += 1
    fileHandle.close()
    return total

# MAIN PROGRAM
wordToMatch = sys.argv[1]
# iterate through all the files
for file in sorted(glob.glob("poems/*.txt")):
    totalWords = findTotal(file)
    totalWordMatches = findWordMatches(file, wordToMatch)
    freq = findWordFrequency(totalWords, totalWordMatches + 1)
    author = file
    author = re.sub(r".txt", "", author)
    author = re.sub(r"_", " ", author)
    author = re.sub(r"poems/", "", author)
    #print (freq)
    #print ("total = %d and matches = %d" % (totalWords, totalWordMatches))
    print ("log((%d+1)/%6d) = %8.4f %s"  % (totalWordMatches, totalWords, math.log(freq), author))

File: lab07/log_probability.pl
#!/usr/bin/perl -w

$wordToSearch = $ARGV[0];

foreach $file (glob "poems/*.txt"){
	$totalWords = findTotalWords($file);
	$wordsMatched = findTotalMatches($file, $wordToSearch);
	$frequency = ($wordsMatched + 1) / $totalWords;
	$logResult = log($frequency);
	$file =~ s/poems\///; # remove the directory
	$file =~ s/_/ /g;
	$file =~ s/\.txt//g;
	printf "log((%d+1)/%6d) = %8.4f %s\n", $wordsMatched, $totalWords, $logResult, $file;
}

# definition of sub functions here

sub findTotalWords{
	my $file = $_[0]; # get the filename from the argument list
	my $total = 0;
	open F, "$file" or die "Failed to open file\n";
	while ($line = <F>){
		my @words = split /[^a-zA-Z]/, $line;
		foreach $word (@words){
			if ($word ne ""){
				$total++;
			}
		}
	}

	return $total;
	#print "$total words\n";
}

sub findTotalMatches{
	my $wordToSearch = lc($_[1]); # get the word from the commandline
	my $filename = $_[0];
	my $total = 0;
	open F, "$filename" or die "Failed to open file for testing\n";
	while ($line = <F>){
		my @words = split /[^a-zA-Z]/, $line;
		foreach $word (@words){
			if (lc($word) eq $wordToSearch){ # the word was found
				$total++;
			}
		}
	}

	return $total;
}
